processing/data_collection/gazette/spiders/to_palmas.py

```diff
@@ -1,75 +1,70 @@
-from dateparser import parse
-import datetime as dt
-import requests
+import datetime
+from urllib.parse import urlencode
+
+import dateparser
 import scrapy
 
 from gazette.items import Gazette
 from gazette.spiders.base import BaseGazetteSpider
 
-last_page_number_xpath = (
-    '//div[@class="paginacao"]/ul[@class="pagination"]/li[last()-1]/a[last()]/text()'
-)
-
 
 class ToPalmasSpider(BaseGazetteSpider):
     TERRITORY_ID = "1721000"
     name = "to_palmas"
-    allowed_domains = ["diariooficial.palmas.to.gov.br", "legislativo.palmas.to.gov.br"]
-    to_palmas_url = (
-        "http://diariooficial.palmas.to.gov.br/todos-diarios/?page={page_number}"
-    )
-    start_urls = ["http://diariooficial.palmas.to.gov.br/todos-diarios/"]
+    allowed_domains = ["diariooficial.palmas.to.gov.br"]
+    start_date = datetime.date(2010, 3, 22)  # First gazette available
+
+    custom_settings = {
+        "MEDIA_ALLOW_REDIRECTS": True,
+    }
+
+    def start_requests(self):
+        dt_inicial = self.start_date.strftime("%d/%m/%Y")
+        dt_final = datetime.date.today().strftime("%d/%m/%Y")
+
+        url_params = {
+            "opcao": "datas",
+            "dt_inicial": dt_inicial,
+            "dt_final": dt_final,
+            "btn_search": "Search",
+        }
+        params = urlencode(url_params)
+        url = f"http://diariooficial.palmas.to.gov.br/resultado-pesquisa/?{params}"
+        yield scrapy.Request(url)
 
     def parse(self, response):
-        last_page_number_str = response.xpath(last_page_number_xpath).extract_first()
-        last_page_number = int(last_page_number_str)
-        for page_number in range(1, last_page_number + 1):
-            url = self.to_palmas_url.format(page_number=page_number)
-            yield scrapy.Request(url=url, callback=self.parse_page)
+        gazettes = response.css(".diario-resultado-pesquisa tbody tr")
+        for gazette in gazettes:
+            gazette_date = gazette.xpath("./td[2]/text()").get()
+            gazette_url = response.urljoin(gazette.css("a::attr(href)").get())
+
+            is_extra_edition = bool(gazette.xpath(".//*[contains(., 'Suplemento')]"))
+            if is_extra_edition:
+                parent_gazette_class_id = gazette.re_first(r"treegrid-parent-(\d+)")
+                parent_gazette_class = f"treegrid-{parent_gazette_class_id}"
+                main_edition_gazette = response.css(f".{parent_gazette_class}")
+                gazette_date = main_edition_gazette.xpath("./td[2]/text()").get()
```

Comment:
What do you think of making this into a separate method like `date_from_previous_row`?

Reply:
Date is not from the previous row, but from the parent edition. We can have more than one extra edition in one day. I splitted as a new method so it is clearer from where we are getting that date.

Reply:
Didn't know that. Thanks :)